src/torchsearchsorted/searchsorted.py

```diff
@@ -1,53 +1,60 @@
+import warnings
 from typing import Optional
 
 import torch
 
-# trying to import the CPU searchsorted
-SEARCHSORTED_CPU_AVAILABLE = True
-try:
-    from torchsearchsorted.cpu import searchsorted_cpu_wrapper
-except ImportError:
-    SEARCHSORTED_CPU_AVAILABLE = False
+from torchsearchsorted.cpu import searchsorted_cpu_wrapper
 
-# trying to import the CUDA searchsorted
-SEARCHSORTED_GPU_AVAILABLE = True
-try:
-    from torchsearchsorted.cuda import searchsorted_cuda_wrapper
-except ImportError:
-    SEARCHSORTED_GPU_AVAILABLE = False
+if torch.cuda.is_available():
+    try:
+        from torchsearchsorted.cuda import searchsorted_cuda_wrapper
+    except ImportError as e:
+        warnings.warn("PyTorch is installed with CUDA support, but "
+                      "torchsearchsorted for CUDA was not installed, "
+                      "please repeat the installation or avoid passing "
+                      "CUDA tensors to the `searchsorted`.")
 
 
-def searchsorted(a: torch.Tensor, v: torch.Tensor,
+def searchsorted(a: torch.Tensor,
+                 v: torch.Tensor,
                  out: Optional[torch.LongTensor] = None,
                  side='left') -> torch.LongTensor:
-    assert len(a.shape) == 2, "input `a` must be 2-D."
-    assert len(v.shape) == 2, "input `v` mus(t be 2-D."
-    assert (a.shape[0] == v.shape[0]
-            or a.shape[0] == 1
-            or v.shape[0] == 1), ("`a` and `v` must have the same number of "
-                                  "rows or one of them must have only one ")
-    assert a.device == v.device, '`a` and `v` must be on the same device'
-
-    result_shape = (max(a.shape[0], v.shape[0]), v.shape[1])
+    if a.ndimension() != 2:
+        raise ValueError(f"Input `a` must be 2D, got shape {a.shape}")
+    if v.ndimension() != 2:
+        raise ValueError(f"Input `v` must be 2D, got shape {v.shape}")
+    if a.device != v.device:
+        raise ValueError(f"Inputs `a` and `v` must on the same device, "
+                         f"got {a.device} and {v.device}")
+
+    a, v = broadcast_tensors(a, v, dim=0)
+
     if out is not None:
-        assert out.device == a.device, "`out` must be on the same device as `a`"
-        assert out.dtype == torch.long, "out.dtype must be torch.long"
-        assert out.shape == result_shape, ("If the output tensor is provided, "
-                                           "its shape must be correct.")
+        if out.shape != v.shape:
+            raise ValueError(f"Output `out` must have the same shape as `v`, "
+                             f"got {out.shape} and {a.shape}")
```

Comment:
This is not true, the output must have the same shape as: `(max(a.shape[0], v.shape[0]), v.shape[1])`

Reply:
Let's see if I get the broadcasting semantic right. 

Let's use these variables:
```
B = batch dimension, say 300
A = number of columns of a, say 15
V = number of columns of v, say 7
```

- If `a` contains many rows and `v` contains many rows, all values in `v` are searched in the corresponding row of `a`, in this case `a.shape[0]` and `v.shape[0]` must match
- If `a` contains only one row and `v` contains many rows, all values in `v` are searched in the first row of `a`
- If `a` contains many rows and `v` contains only one row, all values of the first row of `v` are searched in the all rows of `a`

This is a summary of the cases:
```
   a       v          out
(B, A), (B, V) -> (B, V)
(B, A), (1, V) -> (B, V)
(1, A), (B, V) -> (B, V)
(1, A), (1, V) -> (1, V)
(X, A), (Y, V) -> error
```

When you say:
> the output must have the same shape as `(max(a.shape[0], v.shape[0]), v.shape[1])`
how do you handle the case where `a.shape[0]` and `v.shape[0]` are different and are both > 1? Let's say `max(a.shape[0], v.shape[0]) = max(20, 45) = 45`. In this case how can you match every row of values to be searched to a row of `a`?

A couple of lines before this check, the shapes of `a` and `v` are broadcasted together in the first dimension (raising error if not possible), so I think it's fine to check that `out.shape == v.shape`. The only issue I see is that I put the wrong letter in the exception message:

Instead of `{a.shape}`:
```python
    if out is not None:
        if out.shape != v.shape:
            raise ValueError(f"Output `out` must have the same shape as `v`, "
                             f"got {out.shape} and {a.shape}")
```

Should be `{v.shape}`:
```python
    if out is not None:
        if out.shape != v.shape:
            raise ValueError(f"Output `out` must have the same shape as `v`, "
                             f"got {out.shape} and {v.shape}")
```

Reply:
Hey again. actually, let me be more precise:
your solution works just nice. The only part that I guess doesn't work (am I wrong ?) is in the error message:
after you broadcasted: I don't agree that `out` should be of the same shape as (the origina, i.e. not yet broadcastedl) `v`. Let me give you an example: we may have `v.shape[0]==1` and `a.shape[0]>1`. In this case, we want `out` to be `(a.shape[0], v.shape[1])`.
So I totally agree on the code: at this point (after broadcasting) we may try whether `out` is same shape as `v`. But the user possibly provided another `v`, so the message is not clear. 
See what I mean ?